### ISSUE TITLE
Support proxy-based logins like Authelia, Authentik

### ISSUE THREAD

**Issue by paulgessinger** [paulgessinger/swift-paperless#50]
It's currently not possible to log in to a Paperless-ngx server behind a reverse-proxy login mechanism like Authelia or Authentik.

Possible solutions could be:

1. One possible workaround is disabling the login checks for the `/api` and allow the app to retrieve and use a token bypassing the login service.
2. The Less Paper app uses a [method](https://github.com/paperless-ngx/paperless-ngx/discussions/3283#discussioncomment-7609641) that detects redirects, presents a web view, waits for a login and subsequent login back to Paperless-ngx, and captures cookies in the process. The stored cookies then allow the app to communicate with the `/api` endpoint. This has the possible caveat of requiring the app to throw up a web view where the app has full access to the web content. I'm looking to support this in an opt-in rather than on-by-default way.

Please chime in if I missed anything!

**Comment by Whipbread** [comment]
I for one really like the Less Paper route. Please add this!

**Comment by landmass-deftly-reptile-budget** [comment]
Just for info, if you have this not on the radar:
Version 2.5.0 was released and is now capable to do OIDC
https://github.com/paperless-ngx/paperless-ngx/pull/5190

**Comment by TheRealGramdalf** [comment]
OIDC is by far the preferred method compared to proxy login. I would highly recommend supporting it instead of forward auth, since the latter is more of a hack and has several limitations.

**Comment by leepfrog-ger** [comment]
I've just set up SSO via Keycloak using paperless-ngx 2.5.0, works flawless.
Would love to see that supported in the app!

**Comment by paulgessinger** [comment]
@TheRealGramdalf @leepfrog-ger, so as far as I can tell, this still only allows OIDC **into** paperless, not OAuth2 against Paperless, i.e. Paperless being a provider permitting login via OAuth2 to **it**. 

This means that in this setup, the app still either would have to be able to supply a custom redirect URL (like `swift-paperless://`) to which the access token gets sent, or allow OAuth2 login **into** paperless and then permit access to `/api` via the OAuth2 token rather than the access token.

I might have missed something though.

**Comment by TheRealGramdalf** [comment]
The latter is what typically happens. OIDC (which is *not* the same as Oauth2 - oauth2 is the protocol (e.g. tcp), OIDC is a standard for transmitting data over that protocol (e.g http)
I'm not sure how django-allauth implements tokens exactly, but i do know this much:
- DA keeps a server-local database of all registered accounts (with i.e. postgres), including groups, perms etc
- the OIDC integration allows users to authenticate against an external idp, which then communicates with DA with information about the user *if* they authenticate correctly
- the information about the user could be a uuid, username, display name, or even a list of groups. DA uses this however it sees fit - with the current state of paperless, it just means the user is authenticated (and maybe some username/email info?), group mappings are planned for later.
- once the user is authenticated, I believe DA creates an API token (which is separate from the Oauth2 token) which is then used to allow further requests to the server
- Most implementations I'm aware of open a web browser in a popup window (since you have to go to the idps site to log in), let the authentication flow complete, copy the api token from the header of an outgoing http packet, and then store that in the mobile apps cache for later use.
- as far as URIs go, I'm honestly not certain what part they play - I believe it's so that the app URI (swift-paperless://) can be registered to handle http links for https://paperless.domain.tld (so it can copy the headers), but I'm not really sure. I'm not a mobile dev :p

**Comment by paulgessinger** [comment]
I think what you're describing is correct, but the API token that the identity provider returns to the DA (paperless instance) is the auth token of the provider itself, ie google or authelia or whatever. 

Paperless acts as a client here, and the app is not part of the flow at all, does not receive the upstream auth token, or get any access to the paperless API token at all.

Now if paperless worked as an IDP and the app would log into that, it could maybe use the returned token for further auth, but I don't think that's possible here.

**Comment by paulgessinger** [comment]
Also, if I want to intercept this in the clean way in the app, I would have to supply the redirect URL and initiate the redirect to the IDP. 

But in the current OIDC flow, paperless initiates the redirect, and supplies some url as a callback redirect.

**Comment by beposec** [comment]
The URL recognition for paperless apps with OIDC login could be inspired by the Matrix protocol. With Matrix, the path of the OIDC provider (e.g. https://paperless.example.com/.well-known/paperless/oidc) is located on the domain in a .well-known directory.

https://areweoidcyet.com/client-implementation-guide/

**Comment by paulgessinger** [comment]
Ok I see, that's interesting. 

In this case, the matrix home server delegates to some OIDC provider via well-known, and the app authenticates directly with that OIDC provider, and the home server then (not exactly clear to me how) recognizes that auth as valid. 

I don't think this is implemented in paperless at this time, paperless acts as the client in the OIDC flow. 

On top of this, the OIDC provider has to support dynamic client creation, which I believe neither Authelia nor Authentik currently do.

**Comment by beposec** [comment]
I'm not that deep into the subject. Maybe you don't have to implement the full matrix variant. But maybe you could use the well-known to recognise whether OIDC is active or not and then open the browser with the paperless URL?

**Comment by leepfrog-ger** [comment]
Another app that neatly integrates this and may serve as an example for an own implementation would be Immich with its mobile apps.
It does make use of an app-specific redirect URI with the OIDC provider, see here: https://immich.app/docs/administration/oauth/

I however assume that it would would necessary to modify the web application (in this case paperless-ngx) to support this.

**Comment by paulgessinger** [comment]
I didn't consider the possibility of directly adding the app specific redirect uri to the upstream OIDC provider. 
In that case, I suppose the Immich API also trusts the OIDC token to authenticate the user. 

That's the bit that's missing for Paperless-nix, if I got it right. 

This could definitely work, but I'm not an expert in the backend development.

**Comment by LucaTheHacker** [comment]
I'd still consider the webview method, because that would also allow for people using Proxy providers with Authentik to log in.

For an app like Paperless that may contain sensitive data, I'd definitely prefer to have "two step authentication" rather than hoping that the OAuth2 implementation isn't flawed.

**Comment by xcojonny** [comment]
The Immich app has this kind of feature and it works nicely. 

https://immich.app/docs/overview/quick-start/

**Comment by valentinp72** [comment]
Any news on this subject? For security reasons, I would really like to keep Paperless behind Authelia for 2FA.

**Comment by ocean-haiyang** [comment]
Any updates on this?

**Comment by paulgessinger** [comment]
I think I'm not going to implement forward auto support, but I'm planning to allow you to sign in via OIDC. However, as far as I can tell, that requires some work in the backend, that I haven't had time get into so far. 

Basically, my plan is to add a way for the backend to supply an auth token to the app via redirect, which will then allow me to use a web view to log in.

**Comment by paulgessinger** [comment]
@Whipbread not really. Realistically, maybe I can get something going before the end of the idea.

However, in the upcoming version I'm reworking the login screen, and I'll add the ability to log in with a token directly (by logging in on the web and then pasting the token). That way, it will at least be possible to bypass the OIDC flow and log in anyway.

**Comment by Whipbread** [comment]
> @Whipbread not really. Realistically, maybe I can get something going before the end of the year.
> 
> However, in the upcoming version I'm reworking the login screen, and I'll add the ability to log in with a token directly (by logging in on the web and then pasting the token). That way, it will at least be possible to bypass the OIDC flow and log in anyway.

@paulgessinger 

Thank you for your reply. This is maybe enough for my needs for now. I'll be beta testing 👍

**Comment by paulgessinger** [comment]
@hendrik1120 points out that https://docs.allauth.org/en/latest/headless/index.html exists, which could potentially alleviate the need for a custom token transport mechanism in Paperless-ngx.

**Comment by paulgessinger** [comment]
@hendrik1120 I think the provider token workflow in allauth headless could be what I've been looking for. 
I haven't fully understood if that would require registering the app as a separate client in the OIDC provider next to the actual Paperless-ngx backend.

**Comment by hendrik1120** [comment]
@paulgessinger  Yes, I was also looking at the [Provider Token Endpoint](https://docs.allauth.org/en/latest/headless/openapi-specification/#tag/Authentication:-Providers/paths/~1_allauth~1%7Bclient%7D~1v1~1auth~1provider~1token/post). Since MFA is currently one of the most requested features for paperless using the complete headless api might be also be an option to future-proof the app.

If I remember correctly, the current auth situation in paperless is a complete mess as the Angular frontend still relies on the standard django login with allauth mixed in and authenticates against the drf using the session cookie.

Django headless would be imo the best solution moving forward for both the SPA and all apps. No, it's currently not enabled by default, but can be very easily by adding it to INSTALLED_APPS which is exposed using an ENV and by adding the urls in `urls.py`. As the app will handle the oidc flow itself, it would require as all other apps an `oidc public client`. I will try to dig up some examples.

**Comment by paulgessinger** [comment]
@hendrik1120 I talked to the Paperless-ngx folks, and they don't seem overly interested in implementing this (not enough perceived benefit over existing solution).

Unfortunately, I'm not experienced with Django development, so not sure how effective I'd be at implementing this myself. 

Do you have Django experience? I'd be open to try to collaborate on this.

**Comment by hendrik1120** [comment]
@paulgessinger Sure, on their side it's a two line change. See https://github.com/hendrik1120/paperless-ngx/tree/allauth_headless which enables the _allauth urls. I am currently building you a docker image, but cross building for amd64 takes forever. If you already have a paperless dev containter up and running, it might be easier to just change the two files and restart the container. I will update this post with the image once its ready...

**Comment by hendrik1120** [comment]
Bad news, after a few hours of trying to use the `_allauth/{client}/v1/auth/provider/token`endpoint I ended up with this error message: `Provider does not support token authentication.` Which in my opinion means, that the openid_connect provider as currently implemented in django allauth does not support token authentication. When querying the `_allauth/app/v1/config` there are also no available providers listed. Changing the client to `browser` shows the provider.

Confirmed by: "OpenID support at headless is still a TODO." from [here](https://github.com/pennersr/django-allauth/issues/3960#issuecomment-2221121042). The changes in Paperless required to make this work are still very minimal. The two line change I mentioned before is not quite enough, since this will only provide you with a `session_token`. It is, fully supported to override this behavior to return an api token instead.

I also have a full openid connect flow written in python with PKCE for testing.  I bet that there are much better libraries for swift which will work out of the box.

**Comment by sy6sy2** [comment]
I guess that using the user API token seems to be the right way to do to fix this issue.
My Paperless instance uses Authelia with the OIDC configuration and I was able to create an API token for my Authelia user directly from the Paperless Django admin page. Then I can use this token to connect the Paperparrot iOS on my Paperless instance.

It would be great to have Swift Paperless supporting the "API token" login mode like Paperparrot.
This way, no more problem for Paperless instances protected by OIDC or Forward auth.

**Comment by hendrik1120** [comment]
Sure, using the API token to login allows to login even if the user has been created using OIDC.
That has very litle to do with this issue and is more a feature request to support api token login.

A proper OIDC flow will be much more secure.

**Comment by paulgessinger** [comment]
@sy6sy2 Login with a token only is possible on TestFlight right now, and will be in the next App Store update I push.

**Comment by paulgessinger** [comment]
@hendrik1120 would you be willing to talk through what's needed for OIDC at some point? You seem to have done some thinking and I think it could be useful. I'm open to email / chat / call.

**Comment by sy6sy2** [comment]
> @sy6sy2 Login with a token only is possible on TestFlight right now, and will be in the next App Store update I push.

Oh sorry I didn't find this information on the repo.
I will try it ASAP. Thank you!

**Comment by paulgessinger** [comment]
There's now https://codeberg.org/allauth/django-allauth/issues/4198, and possibly soon a way to run the auth flow with `OIDC` providers in allauth. This would unblock this development.

**Comment by paulgessinger** [comment]
Quick update on this: 

In the meantime, there was an update to allauth that makes this possible. I've been able to run the full OIDC workflow in a python script outside of the app (but exactly how the app would do it) with @hendrik1120's help.

The next step is to try this out in Swift and eventually test this in the app, but I have not made any progress on that yet!

**Comment by m-mattia-m** [comment]
Any updates on this? I’d really love to have this functionality for security purposes. :)

**Comment by svenbaer** [comment]
@paulgessinger Any updates, when this feature will be implemented?

**Comment by vx3r** [comment]
Looking forward to this feature too 👍

**Comment by ternes3** [comment]
@hendrik1120 Sure, would work. Just a hint: https://datatracker.ietf.org/doc/html/rfc8252#section-8.12

**Comment by paulgessinger** [comment]
@hendrik1120 and I think we have a way to do it with proper OIDC directly with the upstream provider and allauth. The only wrinkle ist that in the OIDC client config, the app url must be configured as a valid redirect. 

In my opinion, this is orthogonal to any custom flow that the backend can implement. 

I'll bump a Swift based proof of concept up on my list.

**Comment by paulgessinger** [comment]
I just opened this PR: https://github.com/paulgessinger/swift-paperless/pull/432

**Comment by paulgessinger** [comment]
This is now available on [TestFlight](https://testflight.apple.com/join/bOpOdzwL) in build 174.

**I'll need your help testing this!.** OIDC is added as an additional credential mode after entering the url in the login flow. 

Because the app cannot store a client secret and also cannot get it from the backend, only providers where the app can complete the pkce flow without a secret will work.

That means providers like Google and GitHub do not work, but apparently Microsoft supports this, although I haven't been able to test this. 

 I tested against an Authentik configuration and that works perfect fine.

**Comment by TheRealGramdalf** [comment]
This just seems like a configuration issue to me - making changes on the client/server side won't help, it's up to the oauth2-proxy configuration. IIRC oauth2-proxy has a way to allow certain paths through already. 

In any case if I understand correctly the proper solution is to just get rid of oauth2-proxy entirely for paperless-ngx. Unless you need it for compatibility with some other client, it's just redundant and doesn't add any real security

**Comment by DrissiReda** [comment]
I understand your concerns, that's valid.

**Comment by leepfrog-ger** [comment]
@m-mattia-m while I understand your concern, I think it is better to add that additional layer of security on another level. That could for example be limiting it to internal access only or whitelisted IPs, use some sort of VPN or tailscale/pangolin in front of it, use client certificate authentication or use something like a custom header.

The latter one is the easiest option I believe since you can just check for that custom header at your reverse proxy and let everything through that has it set. And the app supports adding a custom header to all requests.
This also has the advantage that no workarounds or specific additional integration for that double OIDC scenario is necessary (which could introduce undefined behavior or security issues on its own)

**Comment by paulgessinger** [comment]
@alex3305 I discussed the fact that it doesn't show up in the swagger with the Paperless-ngx devs on the pull request. It's possible to include it but it's not trivial.

I'm not entirely sure about the endpoint not responding. Does the app log report anything suspicious when you go to the OIDC mode?

**Comment by alex3305** [comment]
@paulgessinger Yes it did!

<img width="600" height="612" alt="Image" src="https://github.com/user-attachments/assets/4f93de7c-b0d2-4084-9c70-281354b33bb2" />
<br/>
<br/>

That's the logo I have setup in Pocket ID 🙂 .Don't mind the Dutch language 😇 

Logout also works as expected and the same goes for the flow through Safari. For complete reference, see my configuration below.

### Pocket ID

<img width="1666" height="852" alt="Image" src="https://github.com/user-attachments/assets/3e4c49bf-3f96-4a81-9f6f-9c618d4a571c" />

### Paperless-ngx

> [!TIP]
> I use Ansible to deploy Paperless-ngx. Within the `.env` I convert the YAML (above) to inline JSON (below) with a simple [Jinja](https://jinja.palletsprojects.com/) template: `{{ paperless_ngx_social_providers | to_json }}` for maximum readability.

#### `.env`

The basic configuration in my env file, except the `PAPERLESS_SOCIALACCOUNT_PROVIDERS` variable.

```ini
# OAuth / SSO configuration
PAPERLESS_APPS=allauth.socialaccount.providers.openid_connect
PAPERLESS_DISABLE_REGULAR_LOGIN=True
PAPERLESS_REDIRECT_LOGIN_TO_SSO=True
PAPERLESS_SOCIAL_AUTO_SIGNUP=True
PAPERLESS_SOCIALACCOUNT_ALLOW_SIGNUPS=True
```

#### YAML

This is my social account providers configuration that I convert to JSON.

```yaml
paperless_ngx_social_providers:
  openid_connect:
    SCOPE:
      - openid
      - profile
      - email
    OAUTH_PKCE_ENABLED: true
    APPS:
      - provider_id: pocket-id
        name: Pocket ID
        client_id: <REDACTED CLIENT ID>
        settings:
          server_url: "https://id.example.com/.well-known/openid-configuration"
```

#### JSON

And finally the prettified JSON value of the `PAPERLESS_SOCIALACCOUNT_PROVIDERS` parameter.

```json
{
  "openid_connect": {
    "SCOPE": ["openid", "profile", "email"],
    "OAUTH_PKCE_ENABLED": true,
    "APPS": [
      {
        "provider_id": "pocket-id",
        "name": "Pocket ID",
        "client_id": "<REDACTED CLIENT ID>",
        "settings": {
          "server_url": "https://id.example.com/.well-known/openid-configuration"
        }
      }
    ]
  }
}
```

**Comment by paulgessinger** [comment]
I made a proposal for how oidc idps with client secrets could be supported here: https://github.com/paperless-ngx/paperless-ngx/pull/12132

**Comment by DrissiReda** [comment]
What does the closing mean for this issue?

Will you wait for the [allauth PR](https://codeberg.org/allauth/django-allauth/pulls/4681) to be merged?

**Comment by paulgessinger** [comment]
The basic support is there right now. When the changes from that allauth PR make it to Pngx, I'll look into supporting that as well.

**Comment by DrissiReda** [comment]
I think it might be better to leave that PR here as a dependency and keep the ticket open. It's fine to unpin it.

Otherwise people will wonder if this was abandoned

**Comment by paulgessinger** [comment]
https://codeberg.org/allauth/django-allauth/pulls/4741 was just merged in allauth, which would possibly enable broader support for IdPs. I'll investigate

**Comment by mrtncode** [comment]
@paulgessinger Yey 🎉 I am looking forward to it 😃

**Comment by luca910** [comment]
Not sure if this is the right issue for that.
I came to notice that fetching OIDC-Providers from Paperless is not possible when using mTLS altough its configured in the app and works with regular credentials.
I use PocketID as IdP and does not require a client certificate.

<img width="1033" height="254" alt="Image" src="https://github.com/user-attachments/assets/4f79fad0-517e-4651-b8e5-a0bbaba8dd50" />

**Comment by paulgessinger** [comment]
Hi @luca910, it's possible that the OIDC client doesn't send the mTLS cert that's configured before. I didn't think of that. 

I'll look into it, but I'll make a separate issue for it: https://github.com/paulgessinger/swift-paperless/issues/608

**Comment by IngmarStein** [comment]
Is it required to turn off 2fa (TOTP) in Paperless when using OIDC with SwiftPaperless? I‘m trying Paperless-ngx 3.0.0-beta.rc2 with Pocket ID which works on web but with the app, I get
```
Error while logging in
Details: Paperless rejected the OIDC id_token
(HTT 401). {"status": 401, "data": {"flows": [{"id":
"is_pending": true, "types': ["recovery_codes',
"totp"]}]}, "meta': {"is_authenticated': false,
"session_token":
"Xxx"}}
`

**Comment by paulgessinger** [comment]
Hi @IngmarStein I don't think I tested this, so it could be the reason. Is there any reason to want totp on top of OIDC? The app ui doesn't allow for this at the moment in any case.

**Comment by IngmarStein** [comment]
@paulgessinger that could explain it :)
I'm using OIDC as an additional authentication method and I wouldn't want to drop TOTP for the username + password path. Afaict, the setting applies account-wide and is not configurable per authentication method (i.e. I couldn't drop the 2FA requirement just for OIDC). It also works fine on the web.

**Comment by paulgessinger** [comment]
Ok I'll need to look into that then

**Comment by DrissiReda** [comment]
is that the same for passkey auth? I use that and totp as well

**Comment by paulgessinger** [comment]
@DrissiReda does Pngx itself support passkeys? I didn't even know that.

**Comment by DrissiReda** [comment]
it comes from the SSO providers. I use keycloak for example and in web it works if I login via passkeys. not sure if swift paperless can do that via the sso provider

**Comment by gomesp** [comment]
I'm using Passkeys via Pocket-ID. Working like a charm.

**Comment by paulgessinger** [comment]
Right, anything that the IdP does while logging **over there** shouldn't concern the app or Pngx. I use passkeys in Authentik without issues.

**Comment by IngmarStein** [comment]
Yeah, I think we can leave the passkeys topic aside. In my case, I'm also using Pocket ID to authenticate with passkeys and that works fine. The step that currently fails is the TOTP challenge following OIDC, if it's enabled on the account.

**Comment by DrissiReda** [comment]
I see so in this case you mean the TOTP is from paperless local account after the IdP successfully signs you in ?

**Comment by IngmarStein** [comment]
@DrissiReda yes. It's the same flow as for username/password followed by TOTP which SwiftPaperless already handles flawlessly - it's probably just missing for the OIDC method.
